dracut-init: decompress kernel modules and firmware before final compression

MingcongBai opened the pull request:
This pull request changes...

## Changes

Decompress all kernel modules and firmware (if compressed), this allows for better compression ratio as we make the compressed initramfs.

This may save over 10MiB post-XZ.

## Checklist
- [x] I have tested it locally
- [x] I have reviewed and updated any documentation if relevant
- [x] I am providing new code and test(s) for it

MingcongBai committed:
feat: decompress kernel modules and firmware before final compression

Decompress all kernel modules and firmware (if compressed) we put into
the initramfs.  This allows the final compression step to work better.

This may save over 10MiB post-XZ.

Co-authored-by: Mingye Wang <arthur2e5@aosc.io>
Signed-off-by: Mingye Wang <arthur2e5@aosc.io>

MingcongBai commented:
> @MingcongBai Thanks for the PR. This project runs automatic CI test for each PR and for this PR the test is failing. Can you please check ?

Hi, sorry for the delay - I was on new year's vacation here. I will handle this (and my other PR) tomorrow.

MingcongBai commented:
@LaszloGombos Hi there, I was able to eliminate all lint/commit issues but I'm confused as to how the Integration Tests were failing, could you please advise on what to do next?

bdrung commented:
In Debian/Ubuntu we did this (extracting the kernel modules) in initramfs-tools as well, but we changed to putting compressed kernel modules into a separate, uncompressed cpio. See https://discourse.ubuntu.com/t/reduce-initramfs-size-and-speed-up-the-generation-in-ubuntu-23-10/38972

So I suggest doing the same in dracut as well.

Artoria2e5 commented:
> In Debian/Ubuntu we did this (extracting the kernel modules) in initramfs-tools as well, but we changed to putting compressed kernel modules into a separate, uncompressed cpio. See https://discourse.ubuntu.com/t/reduce-initramfs-size-and-speed-up-the-generation-in-ubuntu-23-10/38972
> 
> 
> 
> So I suggest doing the same in dracut as well.

The two approaches do not replace each other. Not recompressing the modules helps generation speed and does not hurt compressed size, but the one thing it does not provide is 10 MB of more compression.

Artoria2e5 commented:
welp. let me run more actions on the local branch and leave the PR one alone.

Artoria2e5 commented:
This is usually a better idea unless the maximum size of the initramfs is somehow the bottleneck. That said now that it's a separate function it would be easy to switch.

I wouldn't want to overload --compress though. Might add a switch for "what to do with compressed kernel modules and firmware" in general, just to leave the "don't compress it" option open as well (it would produce a time saving with basically no size increase)

Artoria2e5 commented:
This is usually a better idea unless the maximum size of the initramfs is somehow the bottleneck. That said now that it's a separate function it would be easy to switch.

I wouldn't want to overload --compress though. Might add a switch for "what to do with compressed kernel modules and firmware" in general, just to leave the "don't compress it" option open as well (it would produce a time saving with basically no size increase)

Nowa-Ammerlaan commented:
See also: https://github.com/dracut-ng/dracut-ng/pull/568

I tried this before but we settled on solving the problem differently by putting the compressed modules in a separate initramfs and cating them together like we do for the microcode.

bdrung commented:
Thanks @Nowa-Ammerlaan for pointing to the previous work. I prefer the approach taken in https://github.com/dracut-ng/dracut-ng/pull/568 to add the decompression to dracut-install. That way one file copy can be skipped. So the initrd generation should be faster.

My recommendation is to pick up https://github.com/dracut-ng/dracut-ng/pull/568 and implement https://github.com/dracut-ng/dracut-ng/pull/568#issuecomment-2341464032 and handle the firmware symbolic links (https://github.com/dracut-ng/dracut-ng/pull/568#issuecomment-2341878027).

Artoria2e5 commented:
All the failures use gzip compression. Probably a clue here, will dig deeper.

The differential shellcheck isn't really my fault -- the existing code looks exactly like that.

Lint issues will be fixed with another substantial update later. Don't want to waste action runs -- I've wasted enough!

Nowa-Ammerlaan commented on dracut-init.sh:
```diff
@@ -1083,6 +1087,86 @@ dracut_kernel_post() {
+
+dracut_kernel_post_split() {
+    ddebug "Splitting kernel modules and firmware into a non-compressed directory ..."
+    mkdir -p "$no_compress_dir"/d/lib/modules/"$kernel"/
```
This might need a check to verify that the modules/firmware actually are compressed.

Artoria2e5 replied:
Yeah, I agree.

bdrung commented:
I'll port the "Put compressed kernel modules and firmware in an uncompressed cpio" patch from initramfs-tools to dracut. So that you do not have to reinvent the wheel for that part.

bdrung committed:
feat: put compressed kernel modules and firmware in an uncompressed cpio

Dracut compresses all files that are put into the initramfs. Kernel
modules and firmware files might already be compressed. Compressing
those files will not give any size savings.

So put compressed kernel modules and firmware files in an uncompressed
cpio instead. This will speed up the initrd generation and extraction
without a significant size change of the initrd.

See
https://salsa.debian.org/kernel-team/initramfs-tools/-/merge_requests/128
and https://lists.ubuntu.com/archives/ubuntu-devel/2023-July/042707.html
for a similar change in initramfs-tools.

Fixes: https://github.com/dracut-ng/dracut-ng/issues/110

Artoria2e5 commented:
Uhm, sure. The Debian approach still needs an `early_cpio` marker to make our `lsinitrd` work independently of `skipcpio`, assuming we don't rewrite the `skipcpio` routine in shell like they do with `unmkinitranfs`.

Artoria2e5 committed:
feat(lsinitrd): add handling of multiple early cpios

I ended up deciding against rewriting the whole thing in terms of
`unmkinitramfs`, mainly because I am not sure about the performance
of pure bash-parsing of cpio.  (Perhaps I worry too much.)

Anyways this not very big change adds support for multiple early
cpios, which lets `lsinitrd` handle the new PR format and the new Debian
format.

Edit 1:  I just realized that some slowness is due to repeatedly
decompressing: getting the version, module list, cmdline etc. all
require decompression, but they do not get the pre_decompress
treatment. Adding it to decompress should help.

Artoria2e5 committed:
chore: add dracut-cpio gitignore entries

Artoria2e5 committed:
fix(dracut-initramfs-restore): add skip-twice support

Artoria2e5 committed:
fix(dracut.sh): restore nul separation

Artoria2e5 committed:
fix(test-40): breakage on multiple libfido2 & lsinitrd on fail

Artoria2e5 committed:
fix(dracut.sh): only find once on noncompressed

Artoria2e5 commented:
Oh come the heck on alpine, what are you doing?

Artoria2e5 committed:
fix(dracut.sh): make busybox find happy

Artoria2e5 commented:
finally!
